src/storage-provider.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+let debugModule = require('debug');
+let assert = require('assert');
+
+/**
+ * Abstract base class for StorageProviders.  Every method of this abstract
+ * base class may be overridden for each provider aside from the constructor.
+ * The constructor should only take a single object as the configuration so
+ * that we can use typed-env-config as the sole system for configuring them.
+ * It is permissible and recommended that implementors validate the config
+ * object.  This class should only deal with storage provider operations and
+ * not any sort of caching or input functions, aside from extra input
+ * validation required by the backend
+ */
+class StorageProvider {
+
+  /**
+   * Store the configuration for this storage provider in the variable
+   * 'this.config' for use in following methods.
+   */
+  constructor(config) {
+    assert(config, 'must pass a configuration object to StorageProvider constructor');
+    assert(config.id, 'all StorageProviders must have an ID');
+    this.id = config.id;
+    this.debug = debugModule(`cloud-mirror:${this.constructor.name}:${this.id}`);
+  }
+
+  /**
+   * Perform any async init code require for this storage provider.  Default action
+   * is for there to be no init code.
+   */
+  async init() { }
+
+  /**
+   * Store a file with the StorageProvider.  This must be overridden with a method
+   * which takes:
+   *   - rawUrl: the original input url as plain text
+   *   - inputStream: a ReadableStream for the object to store
+   *   - headers: HTTP headers required to be set on the object in the storage
+   *     provider
+   *   - storageMetadata: metadata that should be stored in the storage provider
+   *
+   * and inserts the object into the storage provider
+   */
+  async put(rawUrl, inputStream, headers, storageMetadata) {
+    throw new Error('This StorageProvider implementation must implement .put()');
+  }
+
+  /**
+   * Remove an internal address from the storage provider
+   */
+  async purge(internalAddress) {
+    throw new Error('This StorageProvider implementation must implement .purge()');
+  }
+
+  /**
+   * Perform any extra input validation required by the backend
+   */
+  async validateInput(rawUrl) {
+    return true;
+  }
+
+  /**
+   * Take the request-promise response from an HTTP HEAD request for an object
+   * in this storage provider and return a Date object which represents when
+   * this object will be cleaned up from the cache
+   */
+  async expirationDate(response) {
```

Comment:
Instead of doing the `head` request in `CacheManager` I am tempted to suggest we leave it to the `StorageProvider`.
We could call it something like `lookup(rawUrl)` which looks to see if `rawUrl` is cached.

A nice side-effect of this is that some storage providers like azure might be able to extend the expiration time. As to have an object expire 12-36 hours after it's last usage. By updating meta-data if the HEAD request shows _expiration date_ (stored as meta-data) is less than 12 hours away.